pkg/arvo/neo/cod/std/src/fil/axal.txt

```diff
@@ -0,0 +1,338 @@
+# Axal guide
+
+`+axal` is the recursive tree type that Shrubbery uses to build the namespace.
+
+For userspace developers, axals are useful when you're writing `/con` files for your frontend and you want to work with the descendants of whichever shrub you're viewing in the frontend. (e.g. You're rendering a todo item and you want to render its sub-tasks, which are nodes beneath the main task in the namespace.)
+
+We work with axals using the `+of` core defined in `/sur/neo.hoon`, whose arms are documented below.
+
+[note on piths and paths, how `of:neo` is used and how to work with piths and paths, convert from path to pith, etc.]
+
+The `+axal` type itself looks like this.
+
+```
+++  axal
+  |$  [item]
+  [fil=(unit item) kid=(map iota $)]
+```
+
+One way to think of the Shrubbery namespace is as a filesystem, where files could be data, code, or something in between. So the `+axal` represents a directory of such files. The `fil` is the state of the file, which may or may not be null. The `kid` map stores everything below this file in the directory. The map's keys are the path segments immediately below this layer of the directory, and its values are the subtrees that correspond to those paths.
+
+***
+
+## `+anc`
+
+[??? - i see what this is doing just dk why]
+
+```
+++  anc
+  =|  res=(list pith)
+  =|  cur=pith
+  |=  pax=pith
+  ^-  (set pith)
+  ?~  pax
+    (~(gas in *(set pith)) res)
+  =?  res  ?=(^ fil.fat)
+    [cur res]
+  $(fat (~(got by kid.fat) i.pax), pax t.pax, cur (snoc cur i.pax))
+```
+
+???
+
+## `+anc-jab`
+
+[??? - apply gate to nodes?]
+
+```
+++  anc-jab
+  |*  [pax=pith fun=$-(* *)]
+  ^+  fat
+  ?~  pax
+    fat
+  =?  fil.fat  ?=(^ fil.fat)
+    `(fun u.fil.fat)
+  fat(kid (~(put by kid.fat) i.pax $(fat (~(got by kid.fat) i.pax), pax t.pax)))
+```
+
+???
+
+## `+del`
+
+Delete a node.
+
+```
+++  del
+  |=  pax=pith
+  ^+  fat
+  ?~  pax  [~ kid.fat]
+  =/  kid  (~(get by kid.fat) i.pax)
+  ?~  kid  fat
+  fat(kid (~(put by kid.fat) i.pax $(fat u.kid, pax t.pax)))
+```
+
+Returns an axal with the node at path `pax` removed.
+
+## `+dip`
+
+Descend to the axal at this path, [which does what]
```

Comment:
This "zooms in" to the axal at the given path. Given an axal like
```
/foo
/foo/bar
/foo/bar/baz
```
doing `(~(dip of axal) /foo)` would produce
```
/ 
/bar
/bar/baz
```